tests/recorder.test.js

```diff
@@ -0,0 +1,107 @@
+/* global test, expect */
+
+const recorder = require('../libraries/recorder.js');
+const objects = {
+    a: [1, 2, 3, 4],
+    b: [5, 6, 7, 8],
+};
+
+const timeObject = {
+    "1": {
+        "a": [
+            1,
+            2,
+            3,
+            4
+        ],
+        "b": [
+            5,
+            6,
+            7,
+            8
+        ]
+    },
+    "2": {
+        "a": [
+            null,
+            12
+        ]
+    },
+    "3": {
+        "c": [
+            1,
+            2,
+            null,
+            3
+        ]
+    }
+};
+
+test('recorder diff and replay', async () => {
+    recorder.initRecorder(objects);
+    recorder.clearIntervals();
+
+    // Initial state
+    recorder.saveState(0);
+    const expectedTimeObject = {
+        '0': { a: [ 1, 2, 3, 4 ], b: [ 5, 6, 7, 8 ] }
+    };
+    expect(recorder.timeObject).toEqual(expectedTimeObject);
+
+    // 2 -> 12
+    objects.a[1] = 12;
+    recorder.saveState(1);
+
+    expectedTimeObject[1] = { a: [ undefined, 12 ] };
+    expect(recorder.timeObject).toEqual(expectedTimeObject);
+
+    // Delete b
+    delete objects.b;
+    recorder.saveState(2);
+    // No modification, deletions aren't reflected in diff
```

Comment:
Now I realise that we do not get the exact state of var objects at a given time by replay mechanism. In practice, how do we then distinguish whether an object did not change for some time or whether it is already deleted?

Reply:
I believe most cases use an idle timeout type approach "if object not updated in X milliseconds then it's deleted" but this is definitely an area where I don't remember the exact particulars